File: Algorithm/Permutations II.java
public class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        List<List<Integer>> rst = new ArrayList<List<Integer>>();
        if (nums == null || nums.length == 0) {
            return rst;
        }
        helper(rst, nums, 0);
        return rst;
    }
    private void helper(List<List<Integer>> rst, int[] nums, int pos) {
        if (pos == nums.length) {
            rst.add(Arrays.stream(nums).boxed().collect(Collectors.toList()));
            return;
        }
        Set<Integer> set = new HashSet<Integer>();
        for (int i = pos; i < nums.length; ++i) {
            if (set.add(nums[i])) {
                swap(nums, pos, i);
                helper(rst, nums, pos + 1);
                swap(nums, pos, i);
            }
        }
    }
    private void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
}

File: Algorithm/JavaScript/Permutations II.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function(nums) {
    if (!nums.length) {
        return [];
    }
    const rst = [];

    const helper = (current, index) => {
        if (index === current.length) {
            rst.push([...current]);
            return;
        }
        const set = new Set();
        for (let i = index; i < current.length; ++i) {
            if (set.has(current[i])) {
                continue;
            }
            set.add(current[i]);
            [current[index], current[i]] = [current[i], current[index]];
            helper(current, index + 1);
            [current[index], current[i]] = [current[i], current[index]];
        }
    }

    helper(nums, 0);
    return rst;
};

console.log(permuteUnique([0,1,0,0,9]));